### ISSUE TITLE
Pattern matcher does not work with PlusLink

### ISSUE THREAD

**Issue by astroseger** [opencog/atomspace#2284]
The following code return empty list but NumberNode("6") is expected. (code was run inside last opencog/opencog-dev:cli docker)

```scheme
(use-modules (opencog) (opencog exec))
(use-modules (opencog logger))
(NumberNode 1)
(NumberNode 2)
(NumberNode 3)
(NumberNode 4)
(NumberNode 5)
(NumberNode 6)
(define x (GetLink (TypedVariableLink (VariableNode "$X1") (TypeNode "NumberNode")) 
  (EqualLink (PlusLink (VariableNode "$X1") (NumberNode "5")) (NumberNode "11"))))

(cog-execute! x)
```

Or in python even with AndLink it gives the same wrong results
```python
from opencog.type_constructors import *
from opencog.utilities import initialize_opencog
from opencog.bindlink import execute_atom
from opencog.atomspace import TruthValue

# Initialize AtomSpace
atomspace = AtomSpace()
initialize_opencog(atomspace)
summ_11 = GetLink(TypedVariableLink(VariableNode("$X1"), TypeNode("NumberNode")), \
AndLink(EqualLink(PlusLink(VariableNode("$X1"),NumberNode("5")), NumberNode("11"))))
print(execute_atom(atomspace, summ_11)) 
```

**Comment by linas** [comment]
Yes. This is a "known limitation". The EqualLink is an ad-hoc attempt to add support for "Equational Theory" to the pattern matcher  (see https://en.wikipedia.org/wiki/Equational_logic and https://en.wikipedia.org/Equational_theory) 

You were testing something called "arithmetic theory" (see last paragraph of https://en.wikipedia.org/wiki/Theory_(mathematical_logic) )

I have not thought very much about equational theories in general, or what might be needed for better support. This crosses over into Satisfiability Modulo Theories (SMT) https://en.wikipedia.org/wiki/Satisfiability_modulo_theories which is normally considered to be difficult to implement, and it could be interesting to ponder how much of it we could do in the pattern matcher. This remains an open topic.

In the meanwhile, the code should be changed to throw an error for your case. (supporting arithmetic means supporting all of linear algebra; if we also support "greater than", this would require supporting "linear programming" https://en.wikipedia.org/wiki/Linear_programming )

**Comment by linas** [comment]
Oh, I mis-spoke .. since the numbers are already in the atomspace, this should have worked. This is a real bug. Let me see..

**Comment by linas** [comment]
Looking to fix this now.

**Comment by linas** [comment]
Closing, this is fixed by #2286 

Normally, I like to ask newcomers to fix the bugs that they find. However, this bug 
involves a rather complex subsystem, where the fix is only obvious if one already
knows how it works. Basically, it was easier for me to fix, than try to explain to you
how to fix it. :-/
